### PR TITLE
BACK-1542 recipients addresses validation endpoint

### PR DESCRIPTION
**Pull request by mbasire-ledger**
## What is this about?

The workflow of a transaction broadcast includes a preliminary step which consist of validating addresses of the recipients of the transactions. The introduced endpoint allow to validate all the recipients addresses in one call.

**The request** 
`POST /accounts/{uuid}/recipients`
   
  body :  ` ["address1","address2","address3"]`

  

**The answer**
 `{
       "valid": ["address1","address2"]
       "invalid": {
           "address3" : "blablabla"
       }
    }`

## Cute picture of lama
lama

### CODE REVIEW

File: coins/bitcoin/transactor/src/main/scala/co/ledger/lama/bitcoin/transactor/TransactorGrpcService.scala [hunk 1]
```diff
@@ -2,17 +2,31 @@ package co.ledger.lama.bitcoin.transactor
...
   CoinSelectionStrategy,
   FeeLevel,
   PrepareTxOutput,
   RawTransaction
 }
+import co.ledger.lama.bitcoin.transactor.clients.grpc.BitcoinLibClient
+import co.ledger.lama.bitcoin.transactor.protobuf.{
+  ValidateAddressesRequest,
+  ValidateAddressesResponse
+}
```
Comment (estrauser-ledger) [inline coins/bitcoin/transactor/src/main/scala/co/ledger/lama/bitcoin/transactor/TransactorGrpcService.scala:16]: prefix with protobuf to isolate generated case classes

File: coins/bitcoin/transactor/src/main/scala/co/ledger/lama/bitcoin/transactor/TransactorGrpcService.scala [hunk 2]
```diff
@@ -2,17 +2,31 @@ package co.ledger.lama.bitcoin.transactor
...
 import co.ledger.lama.common.utils.UuidUtils
 import com.google.protobuf.ByteString
 import io.grpc.{Metadata, ServerServiceDefinition}
+import co.ledger.lama.bitcoin.common.utils.CoinImplicits._
+import co.ledger.lama.bitcoin.transactor.protobuf.ValidateAddressesResponse.ValidationResult.Result
+import co.ledger.lama.bitcoin.transactor.protobuf.ValidateAddressesResponse.{
+  InvalidAddress,
+  ValidAddress,
+  ValidationResult
+}
```
Comment (estrauser-ledger) [inline coins/bitcoin/transactor/src/main/scala/co/ledger/lama/bitcoin/transactor/TransactorGrpcService.scala:28]: same

File: coins/bitcoin/transactor/src/main/scala/co/ledger/lama/bitcoin/transactor/TransactorGrpcService.scala [hunk 3]
```diff
@@ -125,4 +139,27 @@ class TransactorGrpcService(transactor: Transactor) extends TransactorService wi
...
+      request: ValidateAddressesRequest,
+      ctx: Metadata
+  ): IO[ValidateAddressesResponse] = {
+
+    val network = IO.fromOption(Coin.fromKey(request.coinId).map(_.toNetwork))(
+      new IllegalArgumentException(s"Unknown coin id: ${request.coinId}")
+    )
+
+    def validate(address: String, network: BitcoinNetwork): IO[ValidationResult] =
+      transactor.bitcoinLibClient
```
Comment (estrauser-ledger) [inline coins/bitcoin/transactor/src/main/scala/co/ledger/lama/bitcoin/transactor/TransactorGrpcService.scala:152]: I'd rather have this done directly in the Transactor. I know it needs useless case classes, we should discuss about this.

File: coins/bitcoin/transactor/src/main/scala/co/ledger/lama/bitcoin/transactor/clients/grpc/BitcoinLibClient.scala [hunk 4]
```diff
@@ -33,13 +44,44 @@ trait BitcoinLibClient {
   ): IO[RawTransactionResponse]
 }
 
+object BitcoinLibClient {
+  case class InvalidAddressReason(reason: String) extends AnyVal
+  case class Address(value: String)               extends AnyVal
```
Comment (estrauser-ledger) [inline coins/bitcoin/transactor/src/main/scala/co/ledger/lama/bitcoin/transactor/clients/grpc/BitcoinLibClient.scala:49]: I'd put this in bitcoin.common.models

File: coins/bitcoin/transactor/src/main/scala/co/ledger/lama/bitcoin/transactor/clients/grpc/BitcoinLibClient.scala [hunk 5]
```diff
@@ -33,13 +44,44 @@ trait BitcoinLibClient {
...
+  override def validateAddress(
+      address: Address,
+      network: BitcoinNetwork
+  ): IO[Either[InvalidAddressReason, Address]] = {
+
+    val libGrpcNetwork: libgrpc.BitcoinNetwork = network match {
+      case BitcoinNetwork.MainNet     => libgrpc.BitcoinNetwork.BITCOIN_NETWORK_MAINNET
+      case BitcoinNetwork.TestNet3    => libgrpc.BitcoinNetwork.BITCOIN_NETWORK_TESTNET3
+      case BitcoinNetwork.RegTest     => libgrpc.BitcoinNetwork.BITCOIN_NETWORK_REGTEST
+      case BitcoinNetwork.Unspecified => libgrpc.BitcoinNetwork.BITCOIN_NETWORK_UNSPECIFIED
```
Comment (estrauser-ledger) [inline coins/bitcoin/transactor/src/main/scala/co/ledger/lama/bitcoin/transactor/clients/grpc/BitcoinLibClient.scala:68]: network.toLibGrpxProto 
it's an implicit method declared in transactor.models.implicits

File: coins/bitcoin/transactor/src/main/scala/co/ledger/lama/bitcoin/transactor/clients/grpc/BitcoinLibClient.scala [hunk 6]
```diff
@@ -33,13 +44,44 @@ trait BitcoinLibClient {
...
+          address = address.value,
+          chainParams = Some(ChainParams(ChainParams.Network.BitcoinNetwork(libGrpcNetwork)))
+        ),
+        new Metadata
+      )
+      .map(r =>
+        Either
+          .cond(r.isValid, right = Address(r.address), left = InvalidAddressReason(r.invalidReason))
+      )
+  }
```
Comment (estrauser-ledger) [inline coins/bitcoin/transactor/src/main/scala/co/ledger/lama/bitcoin/transactor/clients/grpc/BitcoinLibClient.scala:83]: I'd rather have a 'fromProto' in the dedicated case classes (I know it's not fun)

File: coins/bitcoin/transactor/src/main/scala/co/ledger/lama/bitcoin/transactor/clients/grpc/BitcoinLibClient.scala (hunk 4 above)
Comment (mbasire-ledger) [reply coins/bitcoin/transactor/src/main/scala/co/ledger/lama/bitcoin/transactor/clients/grpc/BitcoinLibClient.scala:49]: Thoses 2 case classes are part of the contract exposed by the BitcoinLibClient. Only the client of BitcoinLibClient needs it. Why should be move them to a common package ?

File: coins/bitcoin/api/src/main/scala/co/ledger/lama/bitcoin/api/routes/AccountController.scala [hunk 7]
```diff
@@ -402,3 +423,23 @@ object AccountController extends Http4sDsl[IO] with IOLogging {
...
+  implicit val monoid: Monoid[ValidationResult] = new Monoid[ValidationResult] {
+    override def empty: ValidationResult = ValidationResult(List.empty, Map.empty)
+
+    override def combine(x: ValidationResult, y: ValidationResult): ValidationResult =
+      ValidationResult(
+        valid = x.valid ::: y.valid,
+        invalid = (x.invalid.toList ::: y.invalid.toList).toMap
+      )
+  }
+
```
Comment (vbergeron-ledger) [inline coins/bitcoin/api/src/main/scala/co/ledger/lama/bitcoin/api/routes/AccountController.scala:444]: Maybe seed Validated :
https://typelevel.org/cats/datatypes/validated.html

File: coins/bitcoin/transactor/src/main/scala/co/ledger/lama/bitcoin/transactor/clients/grpc/BitcoinLibClient.scala (hunk 5 above)
Comment (mbasire-ledger) [reply coins/bitcoin/transactor/src/main/scala/co/ledger/lama/bitcoin/transactor/clients/grpc/BitcoinLibClient.scala]: ok

**Review by estrauser-ledger — APPROVED** [review]

File: coins/bitcoin/transactor/src/main/scala/co/ledger/lama/bitcoin/transactor/Transactor.scala [hunk 8]
```diff
@@ -55,20 +58,19 @@ class Transactor(
...
+        case Some(custom) => log.info(s"Custom fee: $custom") *> IO.pure(custom)
+        case _            =>
+          // TODO: testnet smart fees is buggy on explorer v3
+          for {
+            smartFee <- coin match {
+              case Coin.BtcTestnet => IO.pure(25642L)
+              case Coin.BtcRegtest => IO.pure(25642L)
+              case c               => explorerClient(c).getSmartFees.map(_.getValue(feeLevel))
+            }
+            _ <- log.info(s"GetSmartFees feeLevel: $feeLevel - feeSatPerKb: $smartFee ")
```
Comment (RomanWlm) [inline coins/bitcoin/transactor/src/main/scala/co/ledger/lama/bitcoin/transactor/Transactor.scala:71]: Log account UID to better tracking

**Review by RomanWlm — CHANGES_REQUESTED** [review]

File: coins/bitcoin/api/src/main/scala/co/ledger/lama/bitcoin/api/routes/AccountController.scala (hunk 7 above)
Comment (mbasire-ledger) [reply coins/bitcoin/api/src/main/scala/co/ledger/lama/bitcoin/api/routes/AccountController.scala:444]: We need to fold all the valid addresses AND the invalid adresses. Validated seems to not fit well in that case.

File: coins/bitcoin/transactor/src/main/scala/co/ledger/lama/bitcoin/transactor/Transactor.scala (hunk 8 above)
Comment (mbasire-ledger) [reply coins/bitcoin/transactor/src/main/scala/co/ledger/lama/bitcoin/transactor/Transactor.scala]: Added. But there is many other log message without an account id. I will not add account id on all messages logged in that  PR. It deserves its own PR.

**Pull request merged**